core/src/main/java/com/linecorp/armeria/internal/common/websocket/WebSocketFrameDecoder.java

```diff
@@ -0,0 +1,403 @@
+                            continue;
+                        }
+                        if (frameOpcode == WebSocketFrameType.PONG.opcode()) {
+                            out.add(WebSocketFrame.ofPooledPong(payloadBuffer));
+                            payloadBuffer = null;
+                            continue;
+                        }
+                        if (frameOpcode == WebSocketFrameType.CLOSE.opcode()) {
+                            receivedClosingHandshake = true;
+                            final Channel channel = ctx.log()
+                                                       .ensureAvailable(RequestLogProperty.SESSION)
+                                                       .channel();
+                            assert channel != null;
+                            closeWebSocketInboundStream(channel);
+                            checkCloseFrameBody(payloadBuffer);
+                            out.add(WebSocketFrame.ofPooledClose(payloadBuffer));
+                            payloadBuffer = null;
+                            webSocketCloseHandler.closeFrameReceived();
+                            continue;
+                        }
+
+                        if (!frameFinalFlag && currentFragmentOpcode == -1) {
+                            currentFragmentOpcode = frameOpcode;
+                        }
+
+                        // Return the frame
+                        if (frameOpcode == WebSocketFrameType.TEXT.opcode()) {
+                            out.add(WebSocketFrame.ofPooledText(payloadBuffer, frameFinalFlag));
+                            payloadBuffer = null;
+                        } else if (frameOpcode == WebSocketFrameType.BINARY.opcode()) {
+                            out.add(WebSocketFrame.ofPooledBinary(payloadBuffer, frameFinalFlag));
+                            payloadBuffer = null;
+                        } else if (frameOpcode == WebSocketFrameType.CONTINUATION.opcode()) {
+                            final boolean isText = currentFragmentOpcode == WebSocketFrameType.TEXT.opcode() ?
+                                                   true : false;
+                            out.add(WebSocketFrame.ofPooledContinuation(payloadBuffer, frameFinalFlag, isText));
+                            payloadBuffer = null;
+                        } else {
+                            protocolViolation(WebSocketCloseStatus.INVALID_MESSAGE_TYPE,
+                                              "Cannot decode web socket frame with opcode: " + frameOpcode);
+                        }
+
+                        if (frameFinalFlag) {
+                            currentFragmentOpcode = -1;
+                        }
+                    } finally {
+                        if (payloadBuffer != null) {
+                            payloadBuffer.release();
+                        }
+                    }
+                    continue;
+                case CORRUPT:
```

Comment:
Is it OK to fall through here?

Reply:
We throw a `WebSocketProtocolViolationException` when it's corrupted. Let me just add a return here. :wink: